### ISSUE TITLE
Seemingly infallible function not triggering unnecesary_wraps

### ISSUE THREAD

**Issue by ejmount** [rust-lang/rust-clippy#14582]
### Summary

I've been looking into a problem I've been having with [rust-fuzz/arbitrary](https://github.com/rust-fuzz/arbitrary) where an error was not being flagged when it seemingly should have been, and noticed what look likes a false negative in clippy in the process. 

https://docs.rs/arbitrary/1.4.1/src/arbitrary/unstructured.rs.html#295-302

### Lint Name

unnecessary_wraps

### Reproducer

Based on the [latest version of the crate](https://docs.rs/arbitrary/1.4.1/src/arbitrary/unstructured.rs.html#295-302) I had this code:

```rust
pub fn int_in_range<T>(&mut self, range: ops::RangeInclusive<T>) -> Result<T>
    where
        T: Int,
    {
        let (result, bytes_consumed) = Self::int_in_range_impl(range, self.data.iter().cloned());

        self.data = &self.data[bytes_consumed..];

        Ok(result)
    }
```

For context, in [the code as published](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=895f391ec7aec044a45a2f224c7e7c2e), `Self::int_in_range_impl` returns `Result`, so the equivalent to the above snippet includes an extra question mark to unwrap it. However, `Self::int_in_range_impl` is *correctly* flagged as `unnecessary_wraps`. When I fixed it to unconditionally return `(T, usize)` and updated callers, I got the above snippet. (Example on playground [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=74861f5a52a8177a8e8b80df5a7787d9
))

I expected to see this happen: a `unneccessary_wraps` warning from clippy, because `int_in_range` seemingly only ever returns `Ok`.

Instead, this happened: no warning

### Version

```text
rustc 1.85.0 (4d91de4e4 2025-02-17)
binary: rustc
commit-hash: 4d91de4e48198da2e33413efdcd9cd2cc0c46688
commit-date: 2025-02-17
host: x86_64-pc-windows-msvc
release: 1.85.0
LLVM version: 19.1.7
```

**Comment by samueltardieu** [comment]
I've edited the labels to make it clearer that this is a documentation issue. Don't hesitate to submit a pull request if you have ideas on how to best formulate this.
